cub/cub/device/device_for.cuh

```diff
@@ -0,0 +1,817 @@
+  OffsetT num_items; // Total number of non-vectorized items
+
+  // TODO Can be extracted into tuning
+  constexpr static int vec_size = 4;
+
+  // Type of the vector that is used to load the input data
+  using vector_t = typename CubVector<T, vec_size>::Type;
+
+  _CCCL_DEVICE void operator()(OffsetT i)
+  {
+    // Surrounding `Bulk` call doesn't invoke this operator on invalid indices, so we don't need to
+    // check for out-of-bounds access here.
+    if (i != partially_filled_vector_id)
+    { // Case of fully filled vector
+      const vector_t vec = *reinterpret_cast<const vector_t*>(input + vec_size * i);
+
+#pragma unroll
+      for (int j = 0; j < vec_size; j++)
+      {
+        op(*(reinterpret_cast<const T*>(&vec) + j));
+      }
+    }
+    else
+    { // Case of partially filled vector
+      for (OffsetT j = i * vec_size; j < num_items; j++)
+      {
+        op(input[j]);
+      }
+    }
+  }
+};
+
+} // namespace for_each
+} // namespace detail
+
+struct DeviceFor
+{
+private:
+  /**
+   * Checks if the pointer is aligned to the given vector type
+   */
+  template <class VectorT, class T>
+  CUB_RUNTIME_FUNCTION static bool is_aligned(const T* ptr)
+  {
+    return (reinterpret_cast<std::size_t>(ptr) & (sizeof(VectorT) - 1)) == 0;
+  }
+
+  template <class InputIteratorT, class OffsetT, class OpT>
+  CUB_RUNTIME_FUNCTION static cudaError_t for_each_n(
+    InputIteratorT first,
+    OffsetT num_items,
+    OpT op,
+    cudaStream_t stream,
+    ::cuda::std::false_type /* do_not_vectorize */)
+  {
+    using wrapped_op_t = detail::for_each::op_wrapper_t<OffsetT, OpT, InputIteratorT>;
+    return detail::for_each::dispatch_t<OffsetT, wrapped_op_t>::dispatch(num_items, wrapped_op_t{first, op}, stream);
+  }
+
+  template <class InputIteratorT, class OffsetT, class OpT>
+  CUB_RUNTIME_FUNCTION static cudaError_t for_each_n(
+    InputIteratorT first, OffsetT num_items, OpT op, cudaStream_t stream, ::cuda::std::true_type /* vectorize */)
+  {
+    auto unwrapped_first = THRUST_NS_QUALIFIER::raw_pointer_cast(&*first);
+    using wrapped_op_t   = detail::for_each::op_wrapper_vectorized_t<OffsetT, OpT, detail::value_t<InputIteratorT>>;
+
+    if (is_aligned<typename wrapped_op_t::vector_t>(unwrapped_first))
+    { // Vectorize loads
+      const OffsetT num_vec_items = cub::DivideAndRoundUp(num_items, wrapped_op_t::vec_size);
+
+      return detail::for_each::dispatch_t<OffsetT, wrapped_op_t>::dispatch(
+        num_vec_items,
+        wrapped_op_t{
+          unwrapped_first, op, num_items % wrapped_op_t::vec_size ? num_vec_items - 1 : num_vec_items, num_items},
+        stream);
+    }
+
+    // Fallback to non-vectorized version
+    return for_each_n(first, num_items, op, stream, ::cuda::std::false_type{});
+  }
+
+public:
+  //! @rst
+  //! Overview
+  //! +++++++++++++++++++++++++++++++++++++++++++++
+  //!
+  //! Applies the function object ``op`` to each index in the provided shape
```

Comment:
I don't like the notion of offset in this context, because it implies offset in something. The shape terminology comes from P2300. The idea behind it is that we might extend the shape to be multidimensional at some point, potentially providing forward progress annotation, so that we could enable shared memory.